x/go/calc/calc_test.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2023 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+package calc_test
+
+import (
+	"go/ast"
+	"go/token"
+	"math"
+	"strconv"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/synnaxlabs/x/calc"
+)
+
+func treesEqual(t1, t2 ast.Expr) bool {
+	switch t1 := t1.(type) {
+	case *ast.BinaryExpr:
+		t2, ok := t2.(*ast.BinaryExpr)
+		if !ok {
+			return false
+		}
+		return t1.Op == t2.Op && treesEqual(t1.X, t2.X) && treesEqual(t1.Y, t2.Y)
+	case *ast.BasicLit:
+		t2, ok := t2.(*ast.BasicLit)
+		if !ok {
+			return false
+		}
+		if t1.Kind != t2.Kind {
+			return false
+		}
+		switch t1.Kind {
+		case token.FLOAT:
+			f1, err := strconv.ParseFloat(t1.Value, 64)
+			if err != nil {
+				return false
+			}
+			f2, err := strconv.ParseFloat(t2.Value, 64)
+			if err != nil {
+				return false
+			}
+			return math.Abs(f1-f2) < 0.000000001
+		}
+	case *ast.Ident:
+		t2, ok := t2.(*ast.Ident)
+		if !ok {
+			return false
+		}
+		return t1.Name == t2.Name
+	}
+	return false
+}
+
+type mockResolver struct {
+	vals map[string]float64
+}
+
+func (m *mockResolver) Resolve(s string) (float64, error) {
+	return m.vals[s], nil
+}
+
+func (m *mockResolver) Set(s string, v float64) error {
+	m.vals[s] = v
+	return nil
+}
+
+var _ = Describe("Calc", func() {
+	Describe("Build", func() {
+		Describe("Build with math expressions", func() {
+			It("Should build 1+1", func() {
+				e := calc.Expression{}
+				err := e.Build("1 + 1")
+				Expect(err).ToNot(HaveOccurred())
+				expectedTree := &ast.BinaryExpr{X: &ast.BasicLit{Kind: token.FLOAT, Value: "1"}, Op: token.ADD, Y: &ast.BasicLit{Kind: token.FLOAT, Value: "1"}}
+				actualTree := e.GetTree().(*ast.BinaryExpr)
+				Expect(treesEqual(expectedTree, actualTree)).To(BeTrue())
+			})
+			It("Should build 1+2*4", func() {
+				e := calc.Expression{}
+				err := e.Build("1 + 2 * 4")
+				Expect(err).ToNot(HaveOccurred())
+				expectedTree := &ast.BinaryExpr{X: &ast.BasicLit{Kind: token.FLOAT, Value: "1"}, Op: token.ADD, Y: &ast.BinaryExpr{X: &ast.BasicLit{Kind: token.FLOAT, Value: "2"}, Op: token.MUL, Y: &ast.BasicLit{Kind: token.FLOAT, Value: "4"}}}
+				actualTree := e.GetTree().(*ast.BinaryExpr)
```

Comment:
Will `GetTree` always return `*ast.BinaryExpr`. If so, why don't we return the concrete type instead of an interface.

Reply:
The `GetTree` method was created mostly so I could make sure that the trees were correct when writing tests. In an actual application of this library, I feel like the tree should be abstracted away, so the user only needs to interact with an Expression object.

Reply:
Ah understood, then this is OK for now. When things like this happen, however, it sometimes points to a type that has too much responsibility. Perhaps split `Expression` into a `Parser`, and an `Evaluator` and then compose them back into the original `Expression` struct?